src/crypto/CHIPCryptoPALPSA.cpp

```diff
@@ -0,0 +1,1747 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      PSA Crypto API based implementation of CHIP crypto primitives
+ */
+
+#include "CHIPCryptoPAL.h"
+
+#include <lib/core/CHIPEncoding.h>
+#include <lib/core/CHIPSafeCasts.h>
+#include <lib/support/BufferWriter.h>
+#include <lib/support/BytesToHex.h>
+#include <lib/support/CHIPArgParser.hpp>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/SafePointerCast.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#include <psa/crypto.h>
+
+#include <mbedtls/bignum.h>
+#include <mbedtls/ecp.h>
+#include <mbedtls/error.h>
+#include <mbedtls/oid.h>
+#include <mbedtls/x509.h>
+#include <mbedtls/x509_csr.h>
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+#include <mbedtls/x509_crt.h>
+#endif // defined(MBEDTLS_X509_CRT_PARSE_C)
+
+#include <string.h>
+#include <type_traits>
+
+#if !defined(MBEDTLS_PSA_CRYPTO_C)
+#error "PSA crypto API not enabled in mbedTLS config"
+#endif
+
+constexpr size_t kMaxErrorStrLen = 128;
+
+// In mbedTLS 3.0.0 direct access to structure fields was replaced with using MBEDTLS_PRIVATE macro.
+#if (MBEDTLS_VERSION_NUMBER >= 0x03000000)
+#define CHIP_CRYPTO_PAL_PRIVATE(x) MBEDTLS_PRIVATE(x)
+#else
+#define CHIP_CRYPTO_PAL_PRIVATE(x) x
+#endif
+
+#if (MBEDTLS_VERSION_NUMBER >= 0x03000000 && MBEDTLS_VERSION_NUMBER < 0x03010000)
+#define CHIP_CRYPTO_PAL_PRIVATE_X509(x) MBEDTLS_PRIVATE(x)
+#else
+#define CHIP_CRYPTO_PAL_PRIVATE_X509(x) x
+#endif
+
+namespace chip {
+namespace Crypto {
+
+namespace {
+
+void logMbedTLSError(int errorCode)
+{
+    if (errorCode != 0)
+    {
+#if defined(MBEDTLS_ERROR_C)
+        char errorStr[kMaxErrorStrLen];
+        mbedtls_strerror(errorCode, errorStr, sizeof(errorStr));
+        ChipLogError(Crypto, "mbedTLS error: %s", errorStr);
+#else
+        // Error codes defined in 16-bit negative hex numbers. Ease lookup by printing likewise
+        ChipLogError(Crypto, "mbedTLS error: -0x%04X", static_cast<uint16_t>(-errorCode));
+#endif
+    }
+}
+
+void logPsaError(psa_status_t status)
+{
+    if (status != 0)
+    {
+        ChipLogError(Crypto, "PSA error: %d", static_cast<int>(status));
+    }
+}
+
+bool isBufferNonEmpty(const uint8_t * data, size_t data_length)
```

Comment:
looks like a general utility function that can be used everywhere in the SDK.

Reply:
Yeah, but other components tend to use `Span` class for memory regions, so perhaps switching the crypto interface to `Span` would make it more aligned with the rest of the codebase.